cyberark/conjur
# The one that everybody hates

Closes #576.

#### What does this pull request do?

- Updates rubocop and reek config to reflect current code style, and recent changes (so there are no warnings currently, as silenced by *TODO* config files),
- refactors security-scan script for more robust docker configuration,
- runs reek and rubocop in the Jenkins pipeline.

#### What background context can you provide?

This pull request will make it easier to keep our code quality at an excellent level. While I didn't fix extant warnings (even the low hanging fruit which can be automatically fixed by rubocop), current offenses are skipped by todo config files. We should gradually fix them with the goal of removing 'todo' settings.

For making sure your changes to Conjur don't break the rules even before you push, I recommend the excellent [overcommit](https://github.com/brigade/overcommit) tool; the configuration for that has been included in this repository for a while. Consult the project page for details, but tl;dr: `gem install overcommit; overcommit --install; overcommit --sign` will make git automatically check all the changes on commit. (We might consider adding it to Gemfile, as in some rvm/rubyenv/bundler environments it's a bit cumbersome to use, ie. it forces you to commit from the outside environment if your bundler environment excludes overcommit libs. Comments welcome.)

#### Where should the reviewer start?

A good place to focus review is the rubocop.yml file, because heated discussions can be had about the settings there.

#### How should this be manually tested?

Running rubocop and reek in the project dir should work, with all files analyzed and no offenses detected. `ci/docker-{rubocop,reek}` scripts should build necessary docker images and run the tools in docker.

#### Link to build in Jenkins (if appropriate)

https://jenkins.conjur.net/job/cyberark--conjur/job/the-one-that-everybody-hates/

#### Questions:
> Is this explained in documentation?

No, but we might consider updating our dev documentation to bring attention to this.
